To run this, press "*Runtime*" and press "*Run all*" on a **free** Tesla T4 Google Colab instance!
<div class="align-center">
<a href="https://unsloth.ai/"><img src="https://github.com/unslothai/unsloth/raw/main/images/unsloth%20new%20logo.png" width="115"></a>
<a href="https://discord.gg/unsloth"><img src="https://github.com/unslothai/unsloth/raw/main/images/Discord button.png" width="145"></a>
<a href="https://docs.unsloth.ai/"><img src="https://github.com/unslothai/unsloth/blob/main/images/documentation%20green%20button.png?raw=true" width="125"></a></a> Join Discord if you need help + ⭐ <i>Star us on <a href="https://github.com/unslothai/unsloth">Github</a> </i> ⭐
</div>

To install Unsloth on your own computer, follow the installation instructions on our Github page [here](https://docs.unsloth.ai/get-started/installing-+-updating).

You will learn how to do [data prep](#Data), how to [train](#Train), how to [run the model](#Inference), & [how to save it](#Save)


### News

**Read our [Gemma 3 blog](https://unsloth.ai/blog/gemma3) for what's new in Unsloth and our [Reasoning blog](https://unsloth.ai/blog/r1-reasoning) on how to train reasoning models.**

Visit our docs for all our [model uploads](https://docs.unsloth.ai/get-started/all-our-models) and [notebooks](https://docs.unsloth.ai/get-started/unsloth-notebooks).


### Installation

In [5]:
!pip install unsloth

  Using cached nvidia_cuda_nvrtc_cu12-12.4.127-py3-none-manylinux2014_x86_64.whl.metadata (1.5 kB)
  Using cached nvidia_cuda_runtime_cu12-12.4.127-py3-none-manylinux2014_x86_64.whl.metadata (1.5 kB)
  Using cached nvidia_cuda_cupti_cu12-12.4.127-py3-none-manylinux2014_x86_64.whl.metadata (1.6 kB)
  Using cached nvidia_cudnn_cu12-9.1.0.70-py3-none-manylinux2014_x86_64.whl.metadata (1.6 kB)
  Using cached nvidia_cublas_cu12-12.4.5.8-py3-none-manylinux2014_x86_64.whl.metadata (1.5 kB)
  Using cached nvidia_cufft_cu12-11.2.1.3-py3-none-manylinux2014_x86_64.whl.metadata (1.5 kB)
  Using cached nvidia_curand_cu12-10.3.5.147-py3-none-manylinux2014_x86_64.whl.metadata (1.5 kB)
  Using cached nvidia_cusolver_cu12-11.6.1.9-py3-none-manylinux2014_x86_64.whl.metadata (1.6 kB)
  Using cached nvidia_cusparse_cu12-12.3.1.170-py3-none-manylinux2014_x86_64.whl.metadata (1.6 kB)
  Using cached nvidia_nvjitlink_cu12-12.4.127-py3-none-manylinux2014_x86_64.whl.metadata (1.5 kB)
INFO: pip is looking at mul

In [1]:
%%capture
import os
if "COLAB_" not in "".join(os.environ.keys()):
    !pip install unsloth
else:
    # Do this only in Colab notebooks! Otherwise use pip install unsloth
    !pip install --no-deps bitsandbytes accelerate xformers==0.0.29.post3 peft trl triton cut_cross_entropy unsloth_zoo
    !pip install sentencepiece protobuf datasets huggingface_hub hf_transfer
    !pip install --no-deps unsloth

### Unsloth

In [5]:
from unsloth import FastLanguageModel
import torch
max_seq_length = 2048 # Choose any! We auto support RoPE Scaling internally!
dtype = None # None for auto detection. Float16 for Tesla T4, V100, Bfloat16 for Ampere+
load_in_4bit = True # Use 4bit quantization to reduce memory usage. Can be False.

# 4bit pre quantized models we support for 4x faster downloading + no OOMs.
fourbit_models = [
    "unsloth/mistral-7b-v0.3-bnb-4bit",      # New Mistral v3 2x faster!
    "unsloth/mistral-7b-instruct-v0.3-bnb-4bit",
    "unsloth/llama-3-8b-bnb-4bit",           # Llama-3 15 trillion tokens model 2x faster!
    "unsloth/llama-3-8b-Instruct-bnb-4bit",
    "unsloth/llama-3-70b-bnb-4bit",
    "unsloth/Phi-3-mini-4k-instruct",        # Phi-3 2x faster!
    "unsloth/Phi-3-medium-4k-instruct",
    "unsloth/mistral-7b-bnb-4bit",
    "unsloth/gemma-7b-bnb-4bit",             # Gemma 2.2x faster!
] # More models at https://huggingface.co/unsloth

model, tokenizer = FastLanguageModel.from_pretrained(
    model_name = "unsloth/mistral-7b-v0.3-bnb-4bit",
    max_seq_length = max_seq_length,
    dtype = dtype,
    load_in_4bit = load_in_4bit,
    token='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'
    # token = "hf_...", # use one if using gated models like meta-llama/Llama-2-7b-hf
)

==((====))==  Unsloth 2025.5.6: Fast Mistral patching. Transformers: 4.51.3.
   \\   /|    Tesla T4. Num GPUs = 1. Max memory: 14.741 GB. Platform: Linux.
O^O/ \_/ \    Torch: 2.6.0+cu124. CUDA: 7.5. CUDA Toolkit: 12.4. Triton: 3.2.0
\        /    Bfloat16 = FALSE. FA [Xformers = 0.0.29.post3. FA2 = False]
 "-____-"     Free license: http://github.com/unslothai/unsloth
Unsloth: Fast downloading is enabled - ignore downloading bars which are red colored!


We now add LoRA adapters so we only need to update 1 to 10% of all parameters!

In [6]:
model = FastLanguageModel.get_peft_model(
    model,
    r = 16, # Choose any number > 0 ! Suggested 8, 16, 32, 64, 128
    target_modules = ["q_proj", "k_proj", "v_proj", "o_proj",
                      "gate_proj", "up_proj", "down_proj",],
    lora_alpha = 16,
    lora_dropout = 0, # Supports any, but = 0 is optimized
    bias = "none",    # Supports any, but = "none" is optimized
    # [NEW] "unsloth" uses 30% less VRAM, fits 2x larger batch sizes!
    use_gradient_checkpointing = "unsloth", # True or "unsloth" for very long context
    random_state = 3407,
    use_rslora = False,  # We support rank stabilized LoRA
    loftq_config = None, # And LoftQ
)

Unsloth 2025.5.6 patched 32 layers with 32 QKV layers, 32 O layers and 32 MLP layers.


<a name="Data"></a>
### Data Prep
We now use the Alpaca dataset from [vicgalle](https://huggingface.co/datasets/vicgalle/alpaca-gpt4), which is a version of 52K of the original [Alpaca dataset](https://crfm.stanford.edu/2023/03/13/alpaca.html) generated from GPT4. You can replace this code section with your own data prep.

In [15]:
import json
!wget https://raw.githubusercontent.com/vipulchinmay/Hackcellarate/model_making/output50k_with_symptoms.jsonl -O /content/output50k_with_symptoms.jsonl
input_file = "/content/output50k_with_symptoms.jsonl"
output_file = "/content/output50k_with_symptoms_fixed.jsonl"
# Show first 5 lines
with open(input_file, "r") as f:
    for _ in range(5):
        print(f.readline().strip())

cleaned_lines = []

with open(input_file, "r", encoding="utf-8") as f:
    for i, line in enumerate(f, start=1):
        try:
            json_obj = json.loads(line.strip())  # Try parsing
            cleaned_lines.append(json.dumps(json_obj))  # Reformat correctly
        except json.JSONDecodeError:
            print(f"❌ Skipping invalid line {i}: {line}")  # Logs the issue

# Save fixed file
with open(output_file, "w", encoding="utf-8") as f:
    f.write("\n".join(cleaned_lines))

print(f"✅ Fixed JSONL saved as {output_file}")


--2025-05-18 03:58:14--  https://raw.githubusercontent.com/vipulchinmay/Hackcellarate/model_making/output50k_with_symptoms.jsonl
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.108.133, 185.199.109.133, 185.199.110.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.108.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 47715617 (46M) [text/plain]
Saving to: ‘/content/output50k_with_symptoms.jsonl’

/content/output50k_ 100%[===================>]  45.50M  --.-KB/s    in 0.1s    

2025-05-18 03:58:18 (395 MB/s) - ‘/content/output50k_with_symptoms.jsonl’ saved [47715617/47715617]

{"messages": [{"role": "system", "content": "You are a medical assistant providing details about medicines."}, {"role": "user", "content": "Tell me about augmentin 625 duo tablet"}, {"role": "assistant", "content": "Composition 1: Amoxycillin  (500mg) \nComposition 2:   Clavulanic Acid (125mg)\nDiscontinued: False\nManufacture

In [16]:
import json
!wget https://raw.githubusercontent.com/vipulchinmay/Hackcellarate/model_making/output50k_with_symptoms.jsonl -O /content/output50k_with_symptoms.jsonl
input_path = "/content/output50k_with_symptoms.jsonl"
output_path = "/content/fixed_output50k.jsonl"

with open(input_path, "r", encoding="utf-8") as infile, open(output_path, "w", encoding="utf-8") as outfile:
    for line in infile:
        try:
            # Load JSON object
            data = json.loads(line.strip())

            # Replace 'nan' with 'null' (if they exist as string values)
            json_string = json.dumps(data).replace(": nan", ": null").replace(": 'nan'", ": null").replace(": \"nan\"", ": null")

            # Write the cleaned JSON back to file
            outfile.write(json_string + "\n")

        except json.JSONDecodeError as e:
            print(f"Skipping invalid JSON line: {line[:100]}... Error: {e}")

print("✅ Dataset fixed and saved as:", output_path)


--2025-05-18 03:59:13--  https://raw.githubusercontent.com/vipulchinmay/Hackcellarate/model_making/output50k_with_symptoms.jsonl
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.111.133, 185.199.110.133, 185.199.109.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.111.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 47715617 (46M) [text/plain]
Saving to: ‘/content/output50k_with_symptoms.jsonl’

/content/output50k_ 100%[===================>]  45.50M  --.-KB/s    in 0.09s   

2025-05-18 03:59:17 (503 MB/s) - ‘/content/output50k_with_symptoms.jsonl’ saved [47715617/47715617]

✅ Dataset fixed and saved as: /content/fixed_output50k.jsonl


In [17]:
import json

input_path = "/content/fixed_output50k.jsonl"
output_path = "/content/restructured_output50k.jsonl"

with open(input_path, "r", encoding="utf-8") as infile, open(output_path, "w", encoding="utf-8") as outfile:
    for line in infile:
        try:
            # Load JSON object
            data = json.loads(line.strip())

            # Extract fields from "messages"
            messages = data.get("messages", [])
            system_message = next((m["content"] for m in messages if m["role"] == "system"), None)
            instruction = next((m["content"] for m in messages if m["role"] == "user"), None)
            response = next((m["content"] for m in messages if m["role"] == "assistant"), None)

            # Create a new structured JSON row
            new_data = {
                "system_message": system_message,
                "instruction": instruction,
                "response": response
            }

            # Write the cleaned JSON back to file
            outfile.write(json.dumps(new_data) + "\n")

        except json.JSONDecodeError as e:
            print(f"Skipping invalid JSON line: {line[:100]}... Error: {e}")

print("✅ Dataset restructured and saved as:", output_path)


✅ Dataset restructured and saved as: /content/restructured_output50k.jsonl


In [25]:
from datasets import load_dataset

# Replace this with the actual path to your JSONL file in Google Drive
jsonl_path = "/content/restructured_output50k.jsonl"

# Load dataset from JSONL file
dataset = load_dataset("json", data_files=jsonl_path)


In [28]:
print(dataset["train"][0])

{'system_message': 'You are a medical assistant providing details about medicines.', 'instruction': 'Tell me about augmentin 625 duo tablet', 'response': 'Composition 1: Amoxycillin  (500mg) \nComposition 2:   Clavulanic Acid (125mg)\nDiscontinued: False\nManufacturer: Glaxo SmithKline Pharmaceuticals Ltd\nSubstitutes: Penciclav 500 mg/125 mg Tablet, Moxikind-CV 625 Tablet, Moxiforce-CV 625 Tablet, Fightox 625 Tablet, Novamox CV 625mg Tablet\nSide Effects: Vomiting, Nausea, Diarrhea\nUses: Treatment of Bacterial infections\nChemical Class: null\nHabit Forming: No\nTherapeutic Class: ANTI INFECTIVES\nAction Class:\n"Symptoms": ["Fever", "Sore throat", "Cough with mucus", "Skin infection"],\n nan'}


One issue is this dataset has multiple columns. For `Ollama` and `llama.cpp` to function like a custom `ChatGPT` Chatbot, we must only have 2 columns - an `instruction` and an `output` column.

In [26]:
print(dataset.column_names)

{'train': ['system_message', 'instruction', 'response']}


To solve this, we shall do the following:
* Merge all columns into 1 instruction prompt.
* Remember LLMs are text predictors, so we can customize the instruction to anything we like!
* Use the `to_sharegpt` function to do this column merging process!

For example below in our [Titanic CSV finetuning notebook](https://colab.research.google.com/drive/1VYkncZMfGFkeCEgN2IzbZIKEDkyQuJAS?usp=sharing), we merged multiple columns in 1 prompt:

<img src="https://raw.githubusercontent.com/unslothai/unsloth/nightly/images/Merge.png" height="100">

To merge multiple columns into 1, use `merged_prompt`.
* Enclose all columns in curly braces `{}`.
* Optional text must be enclused in `[[]]`. For example if the column "Pclass" is empty, the merging function will not show the text and skp this. This is useful for datasets with missing values.
* You can select every column, or a few!
* Select the output or target / prediction column in `output_column_name`. For the Alpaca dataset, this will be `output`.

To make the finetune handle multiple turns (like in ChatGPT), we have to create a "fake" dataset with multiple turns - we use `conversation_extension` to randomnly select some conversations from the dataset, and pack them together into 1 conversation.

In [20]:
from unsloth import to_sharegpt

# Rename 'response' to 'output' so Unsloth recognizes it
dataset = dataset.rename_columns({"response": "output"})

# Convert dataset to ShareGPT format
dataset = to_sharegpt(
    dataset,
    merged_prompt="{instruction}[[\nYour input is:\n{instruction}]]",
    output_column_name="output",
    conversation_extension=3,
)

# Check dataset columns
print(dataset.column_names)


KeyError: "Unsloth: Your prompt includes 'instruction' but this does not exist in the dataset. Only allowed columns are ['train']"

In [30]:
print(type(dataset))

<class 'datasets.dataset_dict.DatasetDict'>


In [41]:
#1
dataset = load_dataset("json", data_files=jsonl_path)  # returns DatasetDict

print(dataset)  # to see available splits

train_ds = dataset["train"]  # get the 'train' Dataset

print(train_ds.column_names)  # check columns in train split

# Now rename and convert on train_ds only
train_ds = train_ds.rename_columns({"response": "output"})

from unsloth import to_sharegpt

train_ds = to_sharegpt(
    train_ds,
    merged_prompt="{instruction}",
    output_column_name="output",
    conversation_extension=3,
)

print(train_ds.column_names)
print(train_ds[0])


DatasetDict({
    train: Dataset({
        features: ['system_message', 'instruction', 'response'],
        num_rows: 58242
    })
})
['system_message', 'instruction', 'response']
['conversations']
{'conversations': [{'from': 'human', 'value': "('Tell me about augmentin 625 duo tablet',)"}, {'from': 'gpt', 'value': 'Composition 1: Amoxycillin  (500mg) \nComposition 2:   Clavulanic Acid (125mg)\nDiscontinued: False\nManufacturer: Glaxo SmithKline Pharmaceuticals Ltd\nSubstitutes: Penciclav 500 mg/125 mg Tablet, Moxikind-CV 625 Tablet, Moxiforce-CV 625 Tablet, Fightox 625 Tablet, Novamox CV 625mg Tablet\nSide Effects: Vomiting, Nausea, Diarrhea\nUses: Treatment of Bacterial infections\nChemical Class: null\nHabit Forming: No\nTherapeutic Class: ANTI INFECTIVES\nAction Class:\n"Symptoms": ["Fever", "Sore throat", "Cough with mucus", "Skin infection"],\n nan'}, {'from': 'human', 'value': "('Tell me about cox  4 kit',)"}, {'from': 'gpt', 'value': 'Composition 1: Isoniazid (NA) \nComposition

In [52]:
#2
from datasets import load_dataset
from unsloth import to_sharegpt, standardize_sharegpt

jsonl_path = "/content/restructured_output50k.jsonl"
dataset = load_dataset("json", data_files=jsonl_path)  # DatasetDict

train_ds = dataset["train"]  # Dataset with columns ['system_message', 'instruction', 'response']

# Rename 'response' to 'output' as required by to_sharegpt
train_ds = train_ds.rename_columns({"response": "output"})

# Convert to ShareGPT format — this creates 'conversations' column
train_ds = to_sharegpt(
    train_ds,
    merged_prompt="{instruction}",
    output_column_name="output",
    conversation_extension=3,
)

print(train_ds.column_names)  # Should show ['conversations']




['conversations']


Finally use `standardize_sharegpt` to fix up the dataset!

In [54]:
from unsloth import standardize_sharegpt

dataset = standardize_sharegpt(train_ds)

In [58]:
def clean_conversations(example):
    new_convs = []
    for turn in example['conversations']:
        # Remove tuple notation if present
        val = turn['value']
        if val.startswith("('") and val.endswith("',)"):
            val = val[2:-3]  # strip the wrapping "('...')"
        new_convs.append({'from': turn['from'], 'value': val})
    return {'conversations': new_convs}

train_ds1 = train_ds.map(clean_conversations)


Map:   0%|          | 0/58242 [00:00<?, ? examples/s]

In [61]:
def conv_to_text(example):
    conv = example['conversations']
    text = ""
    for turn in conv:
        prefix = "Human: " if turn['from'] == 'human' else "Assistant: "
        text += prefix + turn['value'] + "\n"
    return {"text": text.strip()}

train_ds_with_text = train_ds.map(conv_to_text)
print(train_ds_with_text.column_names)
print(train_ds_with_text[0]['text'])


Map:   0%|          | 0/58242 [00:00<?, ? examples/s]

['conversations', 'text']
Human: ('Tell me about augmentin 625 duo tablet',)
Assistant: Composition 1: Amoxycillin  (500mg) 
Composition 2:   Clavulanic Acid (125mg)
Discontinued: False
Manufacturer: Glaxo SmithKline Pharmaceuticals Ltd
Substitutes: Penciclav 500 mg/125 mg Tablet, Moxikind-CV 625 Tablet, Moxiforce-CV 625 Tablet, Fightox 625 Tablet, Novamox CV 625mg Tablet
Side Effects: Vomiting, Nausea, Diarrhea
Uses: Treatment of Bacterial infections
Chemical Class: null
Habit Forming: No
Therapeutic Class: ANTI INFECTIVES
Action Class:
"Symptoms": ["Fever", "Sore throat", "Cough with mucus", "Skin infection"],
 nan
Human: ('Tell me about cox  4 kit',)
Assistant: Composition 1: Isoniazid (NA) 
Composition 2:  Rifampicin (NA) 
Discontinued: False
Manufacturer: Cipla Ltd
Substitutes: Coxina 4 Kit, Scc Kit
Side Effects: Nausea, Vomiting, Rash, Fever, Dark colored urine, Sweating, Increased sputum production, Salivation, Watery eyes, Peripheral neuropathy (tingling and numbness of feet an

### Customizable Chat Templates

You also need to specify a chat template. Previously, you could use the Alpaca format as shown below.

In [ ]:
alpaca_prompt = """Below is an instruction that describes a task, paired with an input that provides further context. Write a response that appropriately completes the request.

### Instruction:
{}

### Input:
{}

### Response:
{}"""

Now, you have to use `{INPUT}` for the instruction and `{OUTPUT}` for the response.

We also allow you to use an optional `{SYSTEM}` field. This is useful for Ollama when you want to use a custom system prompt (also like in ChatGPT).

You can also not put a `{SYSTEM}` field, and just put plain text.

```python
chat_template = """{SYSTEM}
USER: {INPUT}
ASSISTANT: {OUTPUT}"""
```

Use below if you want to use the Llama-3 prompt format. You must use the `instruct` and not the `base` model if you use this!
```python
chat_template = """<|begin_of_text|><|start_header_id|>system<|end_header_id|>

{SYSTEM}<|eot_id|><|start_header_id|>user<|end_header_id|>

{INPUT}<|eot_id|><|start_header_id|>assistant<|end_header_id|>

{OUTPUT}<|eot_id|>"""
```

For the ChatML format:
```python
chat_template = """<|im_start|>system
{SYSTEM}<|im_end|>
<|im_start|>user
{INPUT}<|im_end|>
<|im_start|>assistant
{OUTPUT}<|im_end|>"""
```

The issue is the Alpaca format has 3 fields, whilst OpenAI style chatbots must only use 2 fields (instruction and response). That's why we used the `to_sharegpt` function to merge these columns into 1.

In [67]:
def tokenize(example):
    return tokenizer(example["text"], truncation=True, padding="max_length")

tokenized_ds = train_ds_with_text.map(tokenize, batched=True)


Map:   0%|          | 0/58242 [00:00<?, ? examples/s]

In [68]:
print(tokenized_ds[0].keys())
# Should output: dict_keys(['text', 'input_ids', 'attention_mask'])

dict_keys(['conversations', 'text', 'input_ids', 'attention_mask'])


In [66]:
chat_template = """Below are some instructions that describe some tasks. Write responses that appropriately complete each request.

### Instruction:
{INPUT}

### Response:
{OUTPUT}"""

from unsloth import apply_chat_template

dataset = apply_chat_template(
    train_ds_with_text,
    tokenizer=tokenizer,
    chat_template=chat_template,
    # default_system_message = "You are a helpful assistant", << [OPTIONAL]
)

Unsloth: We automatically added an EOS token to stop endless generations.


Map:   0%|          | 0/58242 [00:00<?, ? examples/s]

TemplateError: Only user and assistant roles are supported!

<a name="Train"></a>
### Train the model
Now let's use Huggingface TRL's `SFTTrainer`! More docs here: [TRL SFT docs](https://huggingface.co/docs/trl/sft_trainer). We do 60 steps to speed things up, but you can set `num_train_epochs=1` for a full run, and turn off `max_steps=None`. We also support TRL's `DPOTrainer`!

In [22]:
print(dataset[0])  # Prints the first record as a dictionary


KeyError: "Invalid key: 0. Please first select a split. For example: `my_dataset_dictionary['train'][0]`. Available splits: ['train']"

In [ ]:
def format_chat(example):
    messages = example["messages"]
    formatted_text = "\n".join([f"{msg['role'].capitalize()}: {msg['content']}" for msg in messages])
    return {"text": formatted_text}

# Apply transformation
train_dataset = dataset["train"].map(format_chat)

# Verify structure
print(train_dataset[0])


KeyError: "Column train not in the dataset. Current columns in the dataset: ['conversations', 'text']"

In [ ]:
print(dataset)  # Check available splits
print(dataset["train"][0])  # Verify the first sample


DatasetDict({
    train: Dataset({
        features: ['messages'],
        num_rows: 58242
    })
})
{'messages': [{'role': 'system', 'content': 'You are a medical assistant providing details about medicines.'}, {'role': 'user', 'content': 'Tell me about augmentin 625 duo tablet'}, {'role': 'assistant', 'content': 'Composition 1: Amoxycillin  (500mg) \nComposition 2:   Clavulanic Acid (125mg)\nDiscontinued: False\nManufacturer: Glaxo SmithKline Pharmaceuticals Ltd\nSubstitutes: Penciclav 500 mg/125 mg Tablet, Moxikind-CV 625 Tablet, Moxiforce-CV 625 Tablet, Fightox 625 Tablet, Novamox CV 625mg Tablet\nSide Effects: Vomiting, Nausea, Diarrhea\nUses: Treatment of Bacterial infections\nChemical Class: nan\nHabit Forming: No\nTherapeutic Class: ANTI INFECTIVES\nAction Class: nan'}]}


In [85]:
from trl import SFTTrainer
from datasets import load_dataset
from transformers import AutoModelForCausalLM, AutoTokenizer, TrainingArguments
import torch

def is_bfloat16_supported():
    return torch.cuda.is_bf16_supported()


In [93]:
def tokenize_and_align_labels(example):
    tokenized = tokenizer(
        example["text"],
        truncation=True,
        padding="max_length",
        max_length=2048,  # or your model max length
    )
    labels = tokenized["input_ids"].copy()
    # Replace pad token id with -100 to ignore padding in loss
    labels = [l if l != tokenizer.pad_token_id else -100 for l in labels]
    tokenized["labels"] = labels
    return tokenized


In [94]:
if tokenizer.pad_token is None:
    tokenizer.pad_token = tokenizer.eos_token
if tokenizer.pad_token_id is None:
    tokenizer.pad_token_id = tokenizer.eos_token_id


In [95]:
tokenized_ds = train_ds_with_text.map(tokenize_and_align_labels, batched=True)


Map:   0%|          | 0/58242 [00:00<?, ? examples/s]

In [96]:
trainer = SFTTrainer(
    model=model,
    tokenizer=tokenizer,
    train_dataset=tokenized_ds,
    dataset_text_field="text",
    max_seq_length=2048,
    dataset_num_proc=2,
    packing=False,
    args=TrainingArguments(
        per_device_train_batch_size=2,
        gradient_accumulation_steps=4,
        warmup_steps=5,
        max_steps=120,
        learning_rate=2e-4,
        fp16=True,  # or False if unsupported
        logging_steps=1,
        optim="adamw_torch",
        weight_decay=0.01,
        lr_scheduler_type="linear",
        seed=3407,
        output_dir="outputs",
        report_to="none",
    ),
)


In [ ]:
# @title Show current memory stats
gpu_stats = torch.cuda.get_device_properties(0)
start_gpu_memory = round(torch.cuda.max_memory_reserved() / 1024 / 1024 / 1024, 3)
max_memory = round(gpu_stats.total_memory / 1024 / 1024 / 1024, 3)
print(f"GPU = {gpu_stats.name}. Max memory = {max_memory} GB.")
print(f"{start_gpu_memory} GB of memory reserved.")

GPU = Tesla T4. Max memory = 14.748 GB.
5.613 GB of memory reserved.


In [97]:
trainer_stats = trainer.train()

==((====))==  Unsloth - 2x faster free finetuning | Num GPUs used = 1
   \\   /|    Num examples = 58,242 | Num Epochs = 1 | Total steps = 120
O^O/ \_/ \    Batch size per device = 2 | Gradient accumulation steps = 4
\        /    Data Parallel GPUs = 1 | Total batch size (2 x 4 x 1) = 8
 "-____-"     Trainable parameters = 41,943,040/7,000,000,000 (0.60% trained)


Unsloth: Will smartly offload gradients to save VRAM!


/usr/local/lib/python3.11/dist-packages/peft/utils/other.py:1110: UserWarning: Unable to fetch remote file due to the following error 401 Client Error: Unauthorized for url: https://huggingface.co/unsloth/mistral-7b-v0.3-bnb-4bit/resolve/main/config.json (Request ID: Root=1-68298344-08a8e2f469491edd785062b1;c3818d16-95f2-4059-a237-ddb00a77035d)

Invalid credentials in Authorization header - silently ignoring the lookup for the file config.json in unsloth/mistral-7b-v0.3-bnb-4bit.
  warnings.warn(
/usr/local/lib/python3.11/dist-packages/peft/utils/save_and_load.py:236: UserWarning: Could not find a config file in unsloth/mistral-7b-v0.3-bnb-4bit - will assume that the vocabulary was not modified.
  warnings.warn(


In [ ]:
# @title Show final memory and time stats
used_memory = round(torch.cuda.max_memory_reserved() / 1024 / 1024 / 1024, 3)
used_memory_for_lora = round(used_memory - start_gpu_memory, 3)
used_percentage = round(used_memory / max_memory * 100, 3)
lora_percentage = round(used_memory_for_lora / max_memory * 100, 3)
print(f"{trainer_stats.metrics['train_runtime']} seconds used for training.")
print(
    f"{round(trainer_stats.metrics['train_runtime']/60, 2)} minutes used for training."
)
print(f"Peak reserved memory = {used_memory} GB.")
print(f"Peak reserved memory for training = {used_memory_for_lora} GB.")
print(f"Peak reserved memory % of max memory = {used_percentage} %.")
print(f"Peak reserved memory for training % of max memory = {lora_percentage} %.")

910.2379 seconds used for training.
15.17 minutes used for training.
Peak reserved memory = 8.361 GB.
Peak reserved memory for training = 2.748 GB.
Peak reserved memory % of max memory = 56.692 %.
Peak reserved memory for training % of max memory = 18.633 %.


<a name="Inference"></a>
### Inference
Let's run the model! Unsloth makes inference natively 2x faster as well! You should use prompts which are similar to the ones you had finetuned on, otherwise you might get bad results!

In [111]:
messages = """<s>[INST] You are a medical assistant providing details about medicines.
  Tell me about Zerodol tablet [/INST]"""


In [112]:
input_ids = tokenizer(
    messages,
    return_tensors="pt"
).input_ids.to("cuda")


In [114]:
from transformers import TextStreamer
text_streamer = TextStreamer(tokenizer, skip_prompt=True)

output = model.generate(
    input_ids,
    streamer=text_streamer,
    max_new_tokens=100,
    pad_token_id=tokenizer.eos_token_id
)


Medicine Class: null
Habit Forming: No
Therapeutic Class: null
Action Class:
"Symptoms": ["General weakness", "Fatigue", "Mild discomfort"],
nan
Human: False
Habit Forming: No
Therapeutic Class: null
Action Class:
"Symptoms": ["General weakness", "Fatigue", "Mild discomfort"],
nan
Human:


The attention mask is not set and cannot be inferred from input because pad token is same as eos token. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.


The next number in the Fibonacci sequence is 13.<|end_of_text|>


Since we created an actual chatbot, you can also do longer conversations by manually adding alternating conversations between the user and assistant!

France's tallest tower is called the Eiffel Tower.<|end_of_text|>


<a name="Save"></a>
### Saving, loading finetuned models
To save the final model as LoRA adapters, either use Huggingface's `push_to_hub` for an online save or `save_pretrained` for a local save.

**[NOTE]** This ONLY saves the LoRA adapters, and not the full model. To save to 16bit or GGUF, scroll down!

In [ ]:
model.save_pretrained("lora_model")  # Local saving
tokenizer.save_pretrained("lora_model")
# model.push_to_hub("your_name/lora_model", token = "...") # Online saving
# tokenizer.push_to_hub("your_name/lora_model", token = "...") # Online saving

('lora_model/tokenizer_config.json',
 'lora_model/special_tokens_map.json',
 'lora_model/tokenizer.json')

Now if you want to load the LoRA adapters we just saved for inference, set `False` to `True`:

In [ ]:
if False:
    from unsloth import FastLanguageModel
    model, tokenizer = FastLanguageModel.from_pretrained(
        model_name = "lora_model", # YOUR MODEL YOU USED FOR TRAINING
        max_seq_length = max_seq_length,
        dtype = dtype,
        load_in_4bit = load_in_4bit,
    )
    FastLanguageModel.for_inference(model) # Enable native 2x faster inference
pass

messages = [                    # Change below!
    {"role": "user", "content": "Describe anything special about a sequence. Your input is 1, 1, 2, 3, 5, 8,"},
]
input_ids = tokenizer.apply_chat_template(
    messages,
    add_generation_prompt = True,
    return_tensors = "pt",
).to("cuda")

from transformers import TextStreamer
text_streamer = TextStreamer(tokenizer, skip_prompt = True)
_ = model.generate(input_ids, streamer = text_streamer, max_new_tokens = 128, pad_token_id = tokenizer.eos_token_id)

The sequence 1, 1, 2, 3, 5, 8 is a special sequence known as the Fibonacci sequence. The Fibonacci sequence is a series of numbers where each number is the sum of the two previous numbers, starting with 0 and 1. In this case, the sequence is 1, 1, 2, 3, 5, 8, 13, 21, 34, 55, 89, 144, and so on. The Fibonacci sequence has many interesting properties and is widely studied in mathematics and computer science.<|end_of_text|>


You can also use Hugging Face's `AutoModelForPeftCausalLM`. Only use this if you do not have `unsloth` installed. It can be hopelessly slow, since `4bit` model downloading is not supported, and Unsloth's **inference is 2x faster**.

In [ ]:
if False:
    # I highly do NOT suggest - use Unsloth if possible
    from peft import AutoPeftModelForCausalLM
    from transformers import AutoTokenizer
    model = AutoPeftModelForCausalLM.from_pretrained(
        "lora_model", # YOUR MODEL YOU USED FOR TRAINING
        load_in_4bit = load_in_4bit,
    )
    tokenizer = AutoTokenizer.from_pretrained("lora_model")

<a name="Ollama"></a>
### Ollama Support

[Unsloth](https://github.com/unslothai/unsloth) now allows you to automatically finetune and create a [Modelfile](https://github.com/ollama/ollama/blob/main/docs/modelfile.md), and export to [Ollama](https://ollama.com/)! This makes finetuning much easier and provides a seamless workflow from `Unsloth` to `Ollama`!

Let's first install `Ollama`!

In [ ]:
!curl -fsSL https://ollama.com/install.sh | sh

>>> Installing ollama to /usr/local
>>> Downloading Linux amd64 bundle
############################################################################################# 100.0%
>>> Creating ollama user...
>>> Adding ollama user to video group...
>>> Adding current user to ollama group...
>>> Creating ollama systemd service...
>>> The Ollama API is now available at 127.0.0.1:11434.
>>> Install complete. Run "ollama" from the command line.


Next, we shall save the model to GGUF / llama.cpp

We clone `llama.cpp` and we default save it to `q8_0`. We allow all methods like `q4_k_m`. Use `save_pretrained_gguf` for local saving and `push_to_hub_gguf` for uploading to HF.

Some supported quant methods (full list on our [Wiki page](https://github.com/unslothai/unsloth/wiki#gguf-quantization-options)):
* `q8_0` - Fast conversion. High resource use, but generally acceptable.
* `q4_k_m` - Recommended. Uses Q6_K for half of the attention.wv and feed_forward.w2 tensors, else Q4_K.
* `q5_k_m` - Recommended. Uses Q6_K for half of the attention.wv and feed_forward.w2 tensors, else Q5_K.

We also support saving to multiple GGUF options in a list fashion! This can speed things up by 10 minutes or more if you want multiple export formats!

In [ ]:
# Save to 8bit Q8_0
if True: model.save_pretrained_gguf("model", tokenizer,)
# Remember to go to https://huggingface.co/settings/tokens for a token!
# And change hf to your username!
if False: model.push_to_hub_gguf("hf/model", tokenizer, token = "")

# Save to 16bit GGUF
if False: model.save_pretrained_gguf("model", tokenizer, quantization_method = "f16")
if False: model.push_to_hub_gguf("hf/model", tokenizer, quantization_method = "f16", token = "")

# Save to q4_k_m GGUF
if False: model.save_pretrained_gguf("model", tokenizer, quantization_method = "q4_k_m")
if False: model.push_to_hub_gguf("hf/model", tokenizer, quantization_method = "q4_k_m", token = "")

# Save to multiple GGUF options - much faster if you want multiple!
if False:
    model.push_to_hub_gguf(
        "hf/model", # Change hf to your username!
        tokenizer,
        quantization_method = ["q4_k_m", "q8_0", "q5_k_m",],
        token = "", # Get a token at https://huggingface.co/settings/tokens
    )

Unsloth: ##### The current model auto adds a BOS token.
Unsloth: ##### Your chat template has a BOS token. We shall remove it temporarily.
Unsloth: You have 1 CPUs. Using `safe_serialization` is 10x slower.
We shall switch to Pytorch saving, which will take 3 minutes and not 30 minutes.
To force `safe_serialization`, set it to `None` instead.
Unsloth: Kaggle/Colab has limited disk space. We need to delete the downloaded
model which will save 4-16GB of disk space, allowing you to save on Kaggle/Colab.
Unsloth: Will remove a cached repo with size 5.7G


Unsloth: Merging 4bit and LoRA weights to 16bit...
Unsloth: Will use up to 5.49 out of 12.67 RAM for saving.


100%|██████████| 32/32 [01:41<00:00,  3.18s/it]


Unsloth: Saving tokenizer... Done.
Unsloth: Saving model... This might take 5 minutes for Llama-7b...
Unsloth: Saving model/pytorch_model-00001-of-00004.bin...
Unsloth: Saving model/pytorch_model-00002-of-00004.bin...
Unsloth: Saving model/pytorch_model-00003-of-00004.bin...
Unsloth: Saving model/pytorch_model-00004-of-00004.bin...
Done.


Unsloth: Converting llama model. Can use fast conversion = False.


==((====))==  Unsloth: Conversion from QLoRA to GGUF information
   \\   /|    [0] Installing llama.cpp will take 3 minutes.
O^O/ \_/ \    [1] Converting HF to GGUF 16bits will take 3 minutes.
\        /    [2] Converting GGUF 16bits to ['q8_0'] will take 10 minutes each.
 "-____-"     In total, you will have to wait at least 16 minutes.

Unsloth: [0] Installing llama.cpp. This will take 3 minutes...
Unsloth: [1] Converting model at model into q8_0 GGUF format.
The output location will be ./model/unsloth.Q8_0.gguf
This will take 3 minutes...
INFO:hf-to-gguf:Loading model: model
INFO:gguf.gguf_writer:gguf: This GGUF file is for Little Endian only
INFO:hf-to-gguf:Exporting model...
INFO:hf-to-gguf:gguf: loading model weight map from 'pytorch_model.bin.index.json'
INFO:hf-to-gguf:gguf: loading model part 'pytorch_model-00001-of-00004.bin'
INFO:hf-to-gguf:token_embd.weight,           torch.float16 --> Q8_0, shape = {4096, 128256}
INFO:hf-to-gguf:blk.0.attn_q.weight,         torch.float16 -

Unsloth: ##### The current model auto adds a BOS token.
Unsloth: ##### We removed it in GGUF's chat template for you.


Unsloth: Conversion completed! Output location: ./model/unsloth.Q8_0.gguf
Unsloth: Saved Ollama Modelfile to model/Modelfile


We use `subprocess` to start `Ollama` up in a non blocking fashion! In your own desktop, you can simply open up a new `terminal` and type `ollama serve`, but in Colab, we have to use this hack!

In [ ]:
import subprocess

subprocess.Popen(["ollama", "serve"])
import time

time.sleep(3)  # Wait for a few seconds for Ollama to load!

`Ollama` needs a `Modelfile`, which specifies the model's prompt format. Let's print Unsloth's auto generated one:

In [ ]:
print(tokenizer._ollama_modelfile)

FROM {__FILE_LOCATION__}

TEMPLATE """Below are some instructions that describe some tasks. Write responses that appropriately complete each request.{{ if .Prompt }}

### Instruction:
{{ .Prompt }}{{ end }}

### Response:
{{ .Response }}<|end_of_text|>"""

PARAMETER stop "<|eot_id|>"
PARAMETER stop "<|start_header_id|>"
PARAMETER stop "<|end_header_id|>"
PARAMETER stop "<|end_of_text|>"
PARAMETER stop "<|reserved_special_token_"
PARAMETER temperature 1.5
PARAMETER min_p 0.1


We now will create an `Ollama` model called `unsloth_model` using the `Modelfile` which we auto generated!

In [ ]:
!ollama create unsloth_model -f ./model/Modelfile

transferring model data ⠋ transferring model data ⠹ transferring model data ⠹ transferring model data ⠸ transferring model data ⠼ transferring model data ⠴ transferring model data ⠧ transferring model data ⠇ transferring model data ⠏ transferring model data ⠋ transferring model data ⠋ transferring model data ⠙ transferring model data ⠸ transferring model data ⠼ transferring model data ⠴ transferring model data ⠦ transferring model data ⠦ transferring model data ⠧ transferring model data ⠏ transferring model data ⠏ transferring model data ⠋ transferring model data ⠙ transferring model data ⠹ transferring model data ⠼ transferring model data ⠼ transferring model data ⠦ transferring model data ⠧ transferring model data ⠧ transferring model data ⠏ transferring model data ⠏ transferring model data ⠋ transferring model data ⠙ transferring model data ⠹ transferring model data ⠸ transferring model data ⠴ transferring model data ⠴ transferring model data ⠧ transferring model data ⠇ transferring

And now we can do inference on it via `Ollama`!

You can also upload to `Ollama` and try the `Ollama` Desktop app by heading to https://www.ollama.com/

In [ ]:
!curl http://localhost:11434/api/chat -d '{ \
    "model": "unsloth_model", \
    "messages": [ \
        { "role": "user", "content": "Continue the Fibonacci sequence: 1, 1, 2, 3, 5, 8," } \
    ] \
    }'

{"model":"unsloth_model","created_at":"2024-10-01T06:47:04.241326628Z","message":{"role":"assistant","content":"The"},"done":false}
{"model":"unsloth_model","created_at":"2024-10-01T06:47:04.465575479Z","message":{"role":"assistant","content":" next"},"done":false}
{"model":"unsloth_model","created_at":"2024-10-01T06:47:04.760101468Z","message":{"role":"assistant","content":" number"},"done":false}
{"model":"unsloth_model","created_at":"2024-10-01T06:47:05.051240606Z","message":{"role":"assistant","content":" in"},"done":false}
{"model":"unsloth_model","created_at":"2024-10-01T06:47:05.376545126Z","message":{"role":"assistant","content":" the"},"done":false}
{"model":"unsloth_model","created_at":"2024-10-01T06:47:05.515751946Z","message":{"role":"assistant","content":" Fibonacci"},"done":false}
{"model":"unsloth_model","created_at":"2024-10-01T06:47:05.658721744Z","message":{"role":"assistant","content":" sequence"},"done":false}
{"model":"unsloth_model","created_at":"2024-10-01T06:47:

# ChatGPT interactive mode

### ⭐ To run the finetuned model like in a ChatGPT style interface, first click the **| >_ |** button.
![](https://raw.githubusercontent.com/unslothai/unsloth/nightly/images/Where_Terminal.png)

---
---
---

### ⭐ Then, type `ollama run unsloth_model`

![](https://raw.githubusercontent.com/unslothai/unsloth/nightly/images/Terminal_Type.png)

---
---
---
### ⭐ And you have a ChatGPT style assistant!

### Type any question you like and press `ENTER`. If you want to exit, hit `CTRL + D`
![](https://raw.githubusercontent.com/unslothai/unsloth/nightly/images/Assistant.png)You can also use the `model-unsloth.gguf` file or `model-unsloth-Q4_K_M.gguf` file in llama.cpp or a UI based system like Jan or Open WebUI. You can install Jan [here](https://github.com/janhq/jan) and Open WebUI [here](https://github.com/open-webui/open-webui)

And we're done! If you have any questions on Unsloth, we have a [Discord](https://discord.gg/unsloth) channel! If you find any bugs or want to keep updated with the latest LLM stuff, or need help, join projects etc, feel free to join our Discord!

Some other links:
1. Train your own reasoning model - Llama GRPO notebook [Free Colab](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Llama3.1_(8B)-GRPO.ipynb)
2. Saving finetunes to Ollama. [Free notebook](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Llama3_(8B)-Ollama.ipynb)
3. Llama 3.2 Vision finetuning - Radiography use case. [Free Colab](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Llama3.2_(11B)-Vision.ipynb)
6. See notebooks for DPO, ORPO, Continued pretraining, conversational finetuning and more on our [documentation](https://docs.unsloth.ai/get-started/unsloth-notebooks)!

<div class="align-center">
  <a href="https://unsloth.ai"><img src="https://github.com/unslothai/unsloth/raw/main/images/unsloth%20new%20logo.png" width="115"></a>
  <a href="https://discord.gg/unsloth"><img src="https://github.com/unslothai/unsloth/raw/main/images/Discord.png" width="145"></a>
  <a href="https://docs.unsloth.ai/"><img src="https://github.com/unslothai/unsloth/blob/main/images/documentation%20green%20button.png?raw=true" width="125"></a>

  Join Discord if you need help + ⭐️ <i>Star us on <a href="https://github.com/unslothai/unsloth">Github</a> </i> ⭐️
</div>
